crates/kubelet/src/state.rs

```diff
@@ -0,0 +1,106 @@
+//! Used to define a state machine of Pod states.
+use log::info;
+
+// pub mod default;
+#[macro_use]
+pub mod macros;
+
+use crate::pod::Pod;
+use k8s_openapi::api::core::v1::Pod as KubePod;
+use kube::api::{Api, PatchParams};
+
+/// Represents result of state execution and which state to transition to next.
+#[derive(Debug)]
+pub enum Transition<S, E> {
+    /// Advance to next node.
+    Advance(S),
+    /// Transition to error node.
+    Error(E),
+    /// This is a terminal node of the state graph.
+    Complete(anyhow::Result<()>),
+}
+
+#[async_trait::async_trait]
+/// Allow for asyncronous cleanup up of PodState.
+pub trait AsyncDrop {
+    /// Clean up PodState.
+    async fn async_drop(&mut self);
+}
+
+#[async_trait::async_trait]
+/// A trait representing a node in the state graph.
+pub trait State<PodState>: Sync + Send + 'static + std::fmt::Debug {
+    /// Provider supplies method to be executed when in this state.
+    async fn next(
+        &self,
+        pod_state: &mut PodState,
+        pod: &Pod,
+    ) -> anyhow::Result<Transition<Box<dyn State<PodState>>, Box<dyn State<PodState>>>>;
+
+    /// Provider supplies JSON status patch to apply when entering this state.
```

Comment:
Would it make more sense to have the status patch come _after_ the work is performed?

Reply:
In my mental model, it makes sense to patch when entering states. If you patch when exiting states, you need to know which state you are going to right? I think it makes sense for each state to own its own patch.